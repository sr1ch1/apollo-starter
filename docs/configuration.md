# Configuration

The following environment variables are supported for configuring the server:

| Name                 | Description                                                                                                            |
| -------------------- | ---------------------------------------------------------------------------------------------------------------------- |
| PORT                 | Port to listen to                                                                                                      |
| CLOSE_GRACE_DELAY    | Time in ms to give server to shutdown gracefuly                                                                        |
| MAX_EVENT_LOOP_DELAY | Allowed time between events until blocking incoming requests                                                           |
| MAX_HEAP_USED_BYTES  | Allowed usage of heap memory until blocking incoming requests                                                          |
| MAX_RSS_BYTES        | Allowed total memory usage until blocking incoming requests                                                            |
| MAX_REQUEST          | Maximum number of request to make before rate limiting kicks in.                                                       |
| TIME_WINDOW          | Time window to look at when counting the number of requests of a user. Accepted values as described in the ms package. |
