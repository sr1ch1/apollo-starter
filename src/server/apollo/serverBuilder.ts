import { ApolloServer } from '@apollo/server';
import type { ContextValue } from '../configuration/contextFactory';
import { resolvers, typeDefs } from '../../models';
import type { ApolloServerPlugin } from '@apollo/server/src/externalTypes/plugins';

export const build = (plugins: ApolloServerPlugin<ContextValue>[] = []) =>
  new ApolloServer<ContextValue>({
    typeDefs,
    resolvers,
    plugins,
  });
