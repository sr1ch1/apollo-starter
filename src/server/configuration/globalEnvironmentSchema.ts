import { z } from 'zod';
import { environmentSchema as populationEnvSchema } from '../../datasources/population/environmentSchema';
import { asFloat, asInteger } from '../../shared/utils/stringUtils';
import { zx } from '../../shared/utils/zodExtensions';

// supported environments
export const EnvironmentEnumSchema = z.enum(['test', 'development', 'staging', 'production']);

// this is the minimal environment schema the server needs.
const coreEnvironmentSchema = z.object({
  // environment type used by the node application
  nodeEnv: EnvironmentEnumSchema.default(EnvironmentEnumSchema.enum.development),

  // this specifies an IP port and its valid range
  port: z.preprocess(asInteger, z.number().gte(1).lt(65536).default(4000)),

  // this specifies the host to listen to
  host: z.string().default('0.0.0.0'),

  // time between events that may pass before the server refuses new requests
  maxEventLoopDelay: z.preprocess(asInteger, z.number().positive().default(1000)),

  // maximum allowed heap usage in bytes before the server refuses new requests
  maxHeapUsedBytes: z.preprocess(asInteger, z.number().positive().default(1073741824)),

  // maximum allowed total memory usage in bytes before the server refuses new requests
  maxRssBytes: z.preprocess(asInteger, z.number().positive().default(1073741824)),

  // rate limiter: maximum number of request to make before rate limiting kicks in.
  maxRequest: z.preprocess(asInteger, z.number().positive().default(1000)),

  // rate limiter: time window to look at when counting the number of requests of a user
  timeWindow: zx.ms(),

  // number of retries when communicating with other REST services and a request times out
  fetchRetries: z.preprocess(asInteger, z.number().positive().default(5)),

  // fetch factor to be used when communicating with other REST services and a request times out
  // an exponential backoff strategy will be used. starting with minTimeout up to maxTimeout
  // with the given number of retries. factor ensures that the retries are distributed exponentially
  // within this range.
  // This formula will be used:
  // Sum[100*x^k, {k, 0, 4}] = 30 * 1000
  // where k is the number of retries-1 and x is the factor.
  // the result can be calculated here if you need different parameters:
  // https://www.wolframalpha.com/input?i=Sum%5B100*x%5Ek%2C+%7Bk%2C+0%2C+4%7D%5D+%3D+30+*+1000
  fetchFactor: z.preprocess(asFloat, z.number().positive().default(3.8626)),

  // The number of milliseconds before starting the first retry (100ms)
  fetchMinTimeout: z.preprocess(asInteger, z.number().positive().default(100)),

  // The maximum number of milliseconds between two retries (30 seconds)
  fetchMaxTimeout: z.preprocess(
    asInteger,
    z
      .number()
      .positive()
      .default(30 * 1000),
  ),
});

// we merge the domain specific environment schemas here into one global schema.
export const globalEnvironmentSchema = coreEnvironmentSchema.merge(populationEnvSchema);
export type EnvironmentSchema = z.infer<typeof globalEnvironmentSchema>;
