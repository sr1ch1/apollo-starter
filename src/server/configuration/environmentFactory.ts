import * as dotenv from 'dotenv';
import type { z } from 'zod';
import { toCamelCaseKey } from '../../shared/utils/stringUtils';
import { globalEnvironmentSchema } from './globalEnvironmentSchema';

export type Environment = z.infer<typeof globalEnvironmentSchema>;

const loadEnvironment = () => {
  dotenv.config();

  // convert environment keys to camel case for better developer experience
  const env = toCamelCaseKey(process.env);
  return globalEnvironmentSchema.parse(env);
};

export class EnvironmentFactory {
  private static _env: Environment | null = null;

  static create(reload = false): Environment {
    if (!this._env || reload) {
      this._env = loadEnvironment();
    }
    return this._env;
  }
}
