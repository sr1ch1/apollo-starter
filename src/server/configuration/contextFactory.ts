import type { EnvironmentSchema } from './globalEnvironmentSchema';
import { IDataSources, buildDataSources } from '../../datasources';
import type { DataSourceConfig } from '@apollo/datasource-rest/dist/RESTDataSource';
import { EnvironmentFactory } from './environmentFactory';

export interface ContextValue {
  env: EnvironmentSchema;
  dataSources: IDataSources;
}

export const ContextBuilder = (config?: DataSourceConfig) => {
  const create = async (): Promise<ContextValue> => {
    const env = EnvironmentFactory.create();
    return {
      env,
      dataSources: config ? buildDataSources(config) : buildDataSources(),
    };
  };

  return create;
};
