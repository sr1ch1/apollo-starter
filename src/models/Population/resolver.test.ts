/**
 * @group integration
 */
import { build } from '../../server/apollo/serverBuilder';
import gql from 'graphql-tag';
import { ContextBuilder } from '../../server/configuration/contextFactory';
import { FetcherStub } from '@sr1ch1/apollo-fetcher-stub';

const query = gql`
  query GetPopulations {
    populations {
      id
      nation
      population
      year
    }
  }
`;
describe('Population Resolver', () => {
  it('should resolve correctly', async () => {
    const stub = new FetcherStub();
    stub
      .get('https://datausa.io/api/data?drilldowns=Nation&measures=Population')
      .responds()
      .withStatusCode(200)
      .withBody(
        `{
    "data": [
        {
            "ID Nation": "01000US",
            "Nation": "United States",
            "ID Year": 2020,
            "Year": "2020",
            "Population": 326569308,
            "Slug Nation": "united-states"
        },
        {
            "ID Nation": "01000US",
            "Nation": "United States",
            "ID Year": 2019,
            "Year": "2019",
            "Population": 324697795,
            "Slug Nation": "united-states"
        }
    ],
    "source": [
        {
            "measures": [
                "Population"
            ],
            "annotations": {
                "source_name": "Census Bureau",
                "source_description": "The American Community Survey (ACS) is conducted by the US Census and sent to a portion of the population every year.",
                "dataset_name": "ACS 5-year Estimate",
                "dataset_link": "http://www.census.gov/programs-surveys/acs/",
                "table_id": "B01003",
                "topic": "Diversity",
                "subtopic": "Demographics"
            },
            "name": "acs_yg_total_population_5",
            "substitutions": []
        }
    ]
}`,
        'application/json',
      );

    const builder = ContextBuilder(stub.fetcher);
    const contextValue = await builder();
    const testServer = build();
    const response = await testServer.executeOperation({ query }, { contextValue });
    expect(response).toMatchSnapshot();
  });
});
