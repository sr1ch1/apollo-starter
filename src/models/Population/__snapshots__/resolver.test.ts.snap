// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Population Resolver should resolve correctly 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "populations": [
          {
            "id": "01000US_01000US",
            "nation": "United States",
            "population": 326569308,
            "year": 2020,
          },
          {
            "id": "01000US_01000US",
            "nation": "United States",
            "population": 324697795,
            "year": 2019,
          },
        ],
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
